working_copy: implement symlinks on Windows with a helper function

gulbanana opened the pull request:
Fixes #2

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

gulbanana commented:
Alright, we've got some test failures on windows-latest. They pass locally, but presumably the runner is not in developer mode and doesn't run as admin - Windows only supports symlinks if one of those things is true! I'll look into modifying the tests, which were previously #[cfg(unix)], so that they look for either success or a special "this Windows instance does not support this feature" error.

In order to achieve that I'll probably make the symlink abstraction the same on both windows and unix, instead of just having the same name. That way we can get better error messages at runtime when this failure occurs on unconfigured-for-symlinks machines.

gulbanana commented:
The exact requirements for symlinks appear to be:
* winver 14972 or higher
* Developer Mode on
* pass flag SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE (`std` does this)

gulbanana commented:
The test changes were more invasive than I hoped and so I've split them into an extra commit. This has some nasty nested error handling, and probably needs a restructure, but I don't know what patterns we prefer for this sort of thing - feedback appreciated!

yuja commented on lib/src/file_util.rs:
```diff
@@ -134,6 +146,39 @@ pub fn persist_content_addressed_temp_file<P: AsRef<Path>>(
+        symlink_dir(original, link)
+    } else {
+        symlink_file(original, link)
+    }
```
Perhaps, the `original` file kind should be passed by caller (based on the repository data, not on the actual filesystem state.) It's scary to access the `original` path here. The path may point to arbitrary filesystem location.

Oh, does it also mean the `original` path can't be switched between file and directory once the symlink is created?

gulbanana replied:
The way it works on Windows is that the symlink itself is to a directory or to a file. If you change the target to be the other thing, it just stops working until you change it back. Here's an example:
![3k9uqVx 1](https://github.com/martinvonz/jj/assets/1612285/a2ed86ee-ea84-49f9-bd41-dafbd5fe6ca6)

What kind of path-switching do you have in mind?

yuja replied:
> What kind of path-switching do you have in mind?

Directory->file on `jj new REV` for example. `jj` wouldn't recreate the symlink itself so long as the symlink data (the `original` path) is unchanged. If git were modeled after Windows symlinks, directory-symlink and file-symlink would be recorded differently, but it's not.

gulbanana replied:
I see the problem. I'll look at the tree data structures and see what it would take to pass the information through.

gulbanana replied:
I looked into this and, as you say, git doesn't store the information we'd need. It just has a relative path, and so that's all `MaterializedTreeValue::Symlink` contains; `TreeState::write_symlink` just receives an absolute and a relative path. We can still implement this, but it's a bit janky: 

`TreeState` knows it's writing a symlink of unknown valence, so it can determine whether it's intended to refer to a directory using `std::path::Path`, and pass this information to the low-level `try_symlink`. Unix can just ignore the flag. Windows has to test whether a symlink of the opposite type already exists, and if so, remove it; then, continue as before.

I've tested this approach and it works in practice. In some future where the backend has sufficiently rich information, we could extract the information from there instead of heuristically determining it in TreeState.

ilyagr replied:
> That's kind of true, but the use of Path::is_dir() has problems...

This feels like a Windows problem.  I wonder if it's feasible for jj to just create a "file" type symlink whenever it's unsure. We could do that if the link points outside the repo, as well.

For example, does something terrible happen on Windows if you create a symlink to a file and then replace the target file with a dir? Even if it does not work any longer, we can argue it's a Windows limitation we can't handle.

yuja replied:
> This feels like a Windows problem. I wonder if it's feasible for jj to just create a "file" type symlink whenever it's unsure. We could do that if the link points outside the repo, as well.

Yeah, I think that's pretty okay. I also think using normal file is good as the minimal support. It will avoid path separator `/` vs `\\` issues and weird symlinks like `///`.

gulbanana replied:
> For example, does something terrible happen on Windows if you create a symlink to a file and then replace the target file with a dir? Even if it does not work any longer, we can argue it's a Windows limitation we can't handle.

The symlink stops working, and JJ is unable to materialize or even `op restore`. You can recover from it, but only by deleting the bad symlink externally.

ilyagr commented:
Optionally, it would be nice to update `docs/windows.md` in this commit. This could also be done in a separate PR.

martinvonz commented on lib/src/file_util.rs:
```diff
@@ -134,6 +146,39 @@ pub fn persist_content_addressed_temp_file<P: AsRef<Path>>(
+        symlink_file(original, link)
+    }
+    .map_err(|err| {
+        if let Some(ERROR_PRIVILEGE_NOT_HELD) = err.raw_os_error() {
```
This `if let` statement doesn't seem to bind any value, so can it be written as `if err.raw_os_error() == Some(ERROR_PRIVILEGE_NOT_HELD)`?

gulbanana replied:
Fixed. This was just me running up against the limits of my Rust knowledge - I didn't know about `Option<T>`s' `impl Eq where T : Eq`.

martinvonz commented on lib/tests/test_local_working_copy.rs:
I would prefer to see the tests in the same commit (and the doc update too). Do tests pass after the first commit? If not, they should definitely be squashed into that commit.

gulbanana replied:
They do pass, but only because they're disabled! I've squashed them.

martinvonz commented on lib/tests/test_local_working_copy.rs:
```diff
@@ -198,10 +197,32 @@ fn test_checkout_file_transitions(backend: TestRepoBackend) {
+                Ok(inner_err)
+                    if matches!(*inner_err, SymlinkError::DeveloperModeNotEnabledError) =>
+                {
+                    return
```
Does this mean we skip the entire test case if developer mode is not enabled? I think we still want to check that we can update between symlinks and that snapshotting works without removing the file when developer mode is not enabled.

gulbanana replied:
I've added a function in testlib which can check whether we only failed because of missing symlink support and continue. However, this didn't put things in a sufficiently useful state to actually do any partial testing (see below or in discord for more than that) - so I just made it skip the test after all :(

That's better than the status quo - previously the tests were being unconditionally skipped on windows - but I'm not sure how we can have a partially functional test. The checkout process deliberately fails if it isn't able to produce a tree equivalent to what should be in the working copy (and that's a good thing). I think we'd need to break the tests up into smaller units, rather than trying to integration-test a subsystem with an internal failure.

martinvonz commented on lib/src/local_working_copy.rs:
```diff
@@ -1177,24 +1176,16 @@ impl TreeState {
-                err: err.into(),
-            })?;
-        }
+        let target = PathBuf::from(&target);
+        try_symlink(&target, disk_path).map_err(|err| CheckoutError::Other {
+            message: format!(
+                "Failed to create symlink from {} to {}",
+                disk_path.display(),
+                target.display()
+            ),
+            err: err.into(),
+        })?;
```
Shouldn't we continue to just print a warning if developer mode is not enabled? Otherwise this commit gives the users who don't have developer mode enabled a worse experience, IIUC.

gulbanana replied:
I implemented this, but found out that skipping the symlink doesn't actually make the checkout work - it's too robust :) The status quo is that this fails entirely:
```
ignoring symlink at \\?\C:\Users\banana\Documents\code\jj\docs\cli-reference.md
Internal error: Failed to check out commit ba70a0c722af06485cb47bf5609eddfc60290a48: Failed to stat file \\?\C:\Users\banana\Documents\code\jj\docs\cli-reference.md: The system cannot find the file specified. (os error 2)
```

I've added a case to CheckoutError so that we get a single structured error message, rather than a skip followed by a generic internal error. To make checkout *work* with symlink support disabled, we'd have to do something like Git's solution of encoding them as text files (and even that doesn't give you an actually-functional source tree).

gulbanana commented:
> Optionally, it would be nice to update `docs/windows.md` in this commit. This could also be done in a separate PR.
> 
> You could say:
> 
> > ### Symbolic link support
> > `jj` supports symlinks on Windows under the following conditions:
> > 
> > * winver 14972 or higher (Windows 10+)
> > * "Developer Mode" is on
> > 
> > If those conditions are not satisfied, `jj` will (for now) crash when the working copy contains symlinks.
> 
> It might also be worth adding that, for colocated repos to work, the git option must be set. (Again, this can go in separately if it's a chore)

I've updated those docs, and addressed (I think!) the requested changes.

ilyagr commented on docs/windows.md:
```diff
@@ -62,3 +62,13 @@ Then only use the `~/my-repo` workspace from Linux.
+working copy contains symlinks. 
+
+For colocated repositories, `git` symlink support must also be enabled using the
+config option `core.symlinks=true`.
```
Let's emphasize that this is a Git config option.

gulbanana replied:
Reworded

yuja commented on lib/src/file_util.rs:
```diff
@@ -134,6 +148,47 @@ pub fn persist_content_addressed_temp_file<P: AsRef<Path>>(
+    // link (directory vs file); if so we have to recreate it
+    if is_dir && matches!(symlink_metadata(link.as_ref()), Ok(meta) if !meta.is_dir()) {
+        remove_dir(link.as_ref()).map_err(SymlinkError::InternalError)?;
+    }
```
What if `!is_dir` and directory symlink already exists?

(I don't know if symlink can be overwritten on Windows, so ignore me if that just works.)

gulbanana replied:
It can be overwritten, as long as you're overwriting a link-to-directory with a link-to-directory (and so on for files).

yuja replied:
then we'll need to handle link-to-dir -> link-to-file case, too?

ah, btw, `symlink()` on Unix fails if the entry already exists. The existing entry should have been deleted by TreeState or LocalWorkingCopy, so we wouldn't have to do `remove_dir()` or `remove_file()`.

yuja commented on lib/src/file_util.rs:
```diff
@@ -15,6 +15,12 @@
+#[cfg(unix)]
+use std::os::unix::fs::symlink;
+#[cfg(windows)]
+use std::os::windows::fs::{symlink_dir, symlink_file};
```
nit: (at your option) these `use` statements can be moved inside the function.

yuja commented on lib/src/file_util.rs:
```diff
@@ -29,6 +35,14 @@ pub struct PathError {
+    #[error("Developer Mode not enabled")]
+    DeveloperModeNotEnabled,
+    #[error("Symlink creation failed")]
+    InternalError(#[source] std::io::Error),
```
nit: `Io(..)` instead of `InternalError(..)`?

gulbanana replied:
Fixed

yuja commented on lib/src/local_working_copy.rs:
```diff
@@ -1177,24 +1176,23 @@ impl TreeState {
-            })?;
-        }
+        let target = PathBuf::from(&target);
+        try_symlink(&target, disk_path, Path::is_dir(disk_path)).map_err(|inner_err| {
```
maybe you mean something like `disk_path.parent().join(target).is_dir()`?

(But see also the comment in the previous discussion https://github.com/martinvonz/jj/pull/2939#discussion_r1481542573)

yuja replied:
Isn't it the function to create the `disk_path` link? Also, the `target` directory/file might not be materialized yet. See https://github.com/martinvonz/jj/pull/2939#discussion_r1481542573.

gulbanana replied:
I think I see the cause of the confusion. I’ve misunderstood Path::is_dir(); I didn’t think it did I/O. My goal was to *calculate* the directory-ness of the path at which the symlink should be created, since (as you say) neither the link nor its target may exist on disk yet; I see now that isn’t possible. 

Here’s a backup plan which I’ll implement in another revision of the branch: check whether the symlink’s target is a repo-relative path; if so, and it’s in the tree, it must be a file. If it’s relative but not in the tree, it’s a directory. If it points outside the tree (can this happen?), we can stat the target. If it points outside the tree but to a file or directory that doesn’t exist… I don’t know. An error saying “please don’t create symlinks to paths outside the repository”??

yuja replied:
> check whether the symlink’s target is a repo-relative path; if so, and it’s in the tree, it must be a file. If it’s relative but not in the tree, it’s a directory. 

nit: if it's a symlink, check recursively (with some cycle detection.)

> If it points outside the tree, we could stat the target.

I would avoid stating external files because it's scary if `jj` read arbitrary unmanaged data (though leaking file node type wouldn't be a security issue.) I think it's okay to fall back to a file symlink (or a normal text file) in that case.

BTW, in the initial implementation, I think it's okay to always assume symlinks as file links. The user can fix it up manually so long as `jj` can read both file and directory symlinks.

gulbanana commented:
I've modified the previous approach based on Martin's patch to create regular files on unsupported platforms. The supportness check goes from a cfg to a flag stored in the TreeState - always true on unix, true on windows if Developer Mode is enabled; turns out we can check that by looking in the registry. The symlink tests are now enabled on all platforms.

I've also addressed the other feedback, with the possible exception of the concern in https://github.com/martinvonz/jj/pull/2939#discussion_r1482755218 - @yuja may want to weigh in.

gulbanana commented:
I think that's everything for now.

martinvonz commented on lib/src/file_util.rs:
```diff
@@ -146,6 +162,51 @@ pub fn persist_content_addressed_temp_file<P: AsRef<Path>>(
     }
 }
 
+/// Symlinks are always available on UNIX, but on Windows they may not be
+/// enabled
```
nit: instead of duplicating this description, i would put only "Symlinks are always available on UNIX" here and then below a short description of when they are available on windows

gulbanana replied:
done

martinvonz commented on lib/src/file_util.rs:
```diff
@@ -146,6 +162,51 @@ pub fn persist_content_addressed_temp_file<P: AsRef<Path>>(
+pub fn check_symlink_support() -> Result<bool, std::io::Error> {
+    let hklm = RegKey::predef(HKEY_LOCAL_MACHINE);
+    let sideloading =
+        hklm.open_subkey("SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\AppModelUnlock")?;
+    let developer_mode: u32 = sideloading.get_value("AllowDevelopmentWithoutDevLicense")?;
```
Will the error message make sense to a user? Maybe we should map it to something more readable (like "Failed to determine if symlinks are supported because <that long key name> wasn't readable"). Actually, none of the callers seem to care about the error, so just return false here instead when we fail to read them instead? How common is it that we would fail to read it?

gulbanana replied:
I don't know of a case when reading the registry entry would fail - you'd have to have some weird broken environment. On the other hand, if the environment is that broken I'd want to know about it!

yuja replied:
> if the environment is that broken I'd want to know about it! I think we should bubble this up like other rare IO-level errors.

Perhaps, tests can be updated to `.unwrap()` the result then? We can also make `TreeState::symlink_support` be evaluated lazily to propagate the error, but I don't know if that'll be worth the added complexity.

gulbanana replied:
the tests are using unwrap() :)

martinvonz commented on lib/src/local_working_copy.rs:
```diff
@@ -1204,28 +1201,23 @@ impl TreeState {
-            symlink(&target, disk_path).map_err(|err| CheckoutError::Other {
+        let target = PathBuf::from(&target);
+        try_symlink(&target, disk_path).map_err(|inner_err| match inner_err {
+            SymlinkError::DeveloperModeNotEnabled => CheckoutError::SymlinkNotSupported(inner_err),
```
This shouldn't happen since we checked the registry, right? I want to avoid errors when updating the working copy whether or not developer mode is enabled. I think it's fine if this only happens if the user had enabled developer mode, then ran `jj new <something>` and disabled developer mode while that was running. But that's such a rare case that I don't think we need a special message for it; just map it to an io::Error.

gulbanana replied:
Originally this error variant existed for tests to use, but the tests can now use check_symlink_support() instead. I've removed it.

martinvonz commented on CHANGELOG.md:
nit: squash this commit into the parent (we would want to cherry-pick or roll them back as one unit)

gulbanana replied:
done

martinvonz commented on docs/windows.md:
```diff
@@ -62,3 +62,13 @@ Then only use the `~/my-repo` workspace from Linux.
+
+## Symbolic link support
+
+`jj` supports symlinks on Windows only when they arere enabled by the operating 
```
nit: typo in "arere"

gulbanana replied:
fixed

martinvonz commented on docs/windows.md:
```diff
@@ -62,3 +62,13 @@ Then only use the `~/my-repo` workspace from Linux.
 [issue-2040]: https://github.com/martinvonz/jj/issues/2040
 [array-op]: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_arrays?view=powershell-7.4#the-array-sub-expression-operator
+
+## Symbolic link support
+
+`jj` supports symlinks on Windows only when they arere enabled by the operating 
+system. This requires Windows 10 version 14972 or higher, as well as Developer
+Mode. If those conditions are not satisfied, `jj` will materialize symlinks as
+ordinary files.
+
+For colocated repositories, Git support must also be enabled using the
+`git config` option `core.symlinks=true`.
```
Thanks for documenting it and remembering to mention colocated repos.

yuja commented on lib/src/local_working_copy.rs:
```diff
@@ -549,6 +548,7 @@ impl TreeState {
             file_states: FileStatesMap::new(),
             sparse_patterns: vec![RepoPathBuf::root()],
             own_mtime: MillisSinceEpoch(0),
+            symlink_support: check_symlink_support().unwrap_or(false),
```
Perhaps, `get_updated_tree_value()` can also be changed to use this flag?
```rust
            let current_tree_values = current_tree.path_value(repo_path);
            let new_file_type = if cfg!(windows) { // XXX HERE
                let mut new_file_type = new_file_state.file_type.clone();
                if matches!(new_file_type, FileType::Normal { .. })
                    && matches!(current_tree_values.as_normal(), Some(TreeValue::Symlink(_)))
                {
                    new_file_type = FileType::Symlink;
                }
                new_file_type
            } else {
                new_file_state.file_type.clone()
            };
```

gulbanana replied:
good catch, i've changed it

yuja commented on lib/src/file_util.rs:
```diff
@@ -146,6 +162,51 @@ pub fn persist_content_addressed_temp_file<P: AsRef<Path>>(
+) -> Result<(), SymlinkError> {
+    symlink(original, link).map_err(SymlinkError::Io)
+}
+
```
nit: since there are multiple platform-specific functions, they can be organized as inner modules (if you prefer):

```rust
pub use self::platform::*;

#[cfg(unix)]
mod platform {
}

#[cfg(windows)]
mod platform {
}
```

gulbanana replied:
sure, i've done that and moved the platform-specific use statements inside the inner modules

gulbanana commented on Cargo.lock:
fixed. i must have done something wrong with a squash

gulbanana commented:
Addressed feedback and rebased.

yuja commented on CHANGELOG.md:
```diff
@@ -73,9 +73,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed bugs
 
-* On Windows, symlinks in the repo are now materialized as regular files in the
+* On Windows, symlinks in the repo are now are now supported when Developer Mode is enabled.
```
nit: double "are now"

gulbanana replied:
cripes, how did i make so many typos... i think i've been been doing something as i learn `jj move` workflows. fixed!

yuja commented on lib/src/file_util.rs:
```diff
@@ -146,6 +148,54 @@ pub fn persist_content_addressed_temp_file<P: AsRef<Path>>(
+        Ok(true)
+    }
+
+    #[cfg(unix)]
```
nit: redundant `#[cfg(unix)]`

gulbanana replied:
removed

yuja commented on lib/src/file_util.rs:
```diff
@@ -146,6 +148,54 @@ pub fn persist_content_addressed_temp_file<P: AsRef<Path>>(
+        Ok(developer_mode == 1)
+    }
+
+    #[cfg(windows)]
```
ditto

gulbanana replied:
also removed

yuja commented on CHANGELOG.md:
In commit message:
>    working_copy: implement symlinks on Windows with a helper function
>    tests: enable symlink tests on windows, ignoring failures due to disabled developer mode
>    docs: add note on Windows symlink support

Maybe not updated after squashing?

gulbanana replied:
That was deliberate, since it changes all three things - but I take it this is not the right style to use!

gulbanana replied:
I've updated the description to the multi-paragraph style that's used in other finished commits

yuja reviewed (approved):
lgtm, thanks.
